### PR TITLE
Warn about unused crate deps

### PR DESCRIPTION
**Pull request by jsgf**
Implements #57274 by adding -Wunused-crate-dependencies. This will warn about any `--extern` option on the command line which isn't referenced by the crate source either via `use` or `extern crate`.

Crates which are added for some side effect but are otherwise unreferenced - such as for symbols they define - the warning can be suppressed with `use somecrate as _;`.

If a crate has multiple aliases (eg using `foo = { package = "bar" }` in `Cargo.toml`), then it will warn about each unused alias.

This does not consider crate added by some other means than `--extern`, including the standard library. It also doesn't consider any crate without `add_prelude` set (though I'm not sure about this).

Unfortunately this probably [does not yet work well with Cargo](https://github.com/rust-lang/rust/issues/57274#issuecomment-624839355) as it will over-specify crates, causing spurious warnings. As a result, this lint is "allow" by default and must be explicitly enabled either via `#![warn(unused_crate_dependencies)]` or with `-Wunused-crate-dependencies`.

### CODE REVIEW

**Comment by jsgf** [comment]
I'll split the deps cleanup into a separate PR.

**Comment by jsgf** [comment]
This seems to work well in practice on a real codebase, except that the emitted diagnostics don't give any clue as to which crate they relate to when they're all jumbled up in a parallel builds. 

@estebank @petrochenkov Since we can't (yet - I have some thoughts) report the actual source location of the dependency, can it/should it report the top-level module source or some other way to identify the crate? Something like `foo/src/lib.rs` (or even `foo/src/lib.rs:1:1` at a pinch).

**Comment by ehuss** [comment]
On option is to do something similar to what other global errors do, like [main not found](https://github.com/rust-lang/rust/blob/42acd9086f93fdc68a61cd79152abca77c6cd369/src/librustc_passes/entry.rs#L171-L197).  Explicitly add the name of the crate to the message, perhaps with a suggestion on how to fix it.  This is similar to what #69203 did.

```
warning: external crate `glob` is unused in crate `foo`
  |
  = help: consider removing `glob` from `Cargo.toml` or adding `extern crate glob as _;` to `src/lib.rs`
```

I think it would be nice to suggest editing `Cargo.toml` only if the "CARGO" environment variable is set.  I've been thinking about doing this for some other errors that could be improved by suggesting edits to `Cargo.toml`.  I'm not sure what the non-Cargo message should say, maybe "remove the `--extern` flag?", seems kinda weird.

**Comment by jsgf** [comment]
@ehuss Yes, just the crate name is better than the full span of the top-level module. As far as I can see the the early buffered lint stuff doesn't support additional help/advice diagnostics, so I can skip that for now.

In a later PR I'd like to investigate an extension to allow Cargo/Buck/etc give rustc something it can put in its diagnostics for each `--extern` so that it can report something sensible in the context of the build environment.

**Comment by estebank** [comment]
> As far as I can see the the early buffered lint stuff doesn't support additional help/advice diagnostics

It's not available in all contexts, but `buffer_lint_with_diagnostic` exists which takes an additional `enum` that can carry extra information for the lints that can then be customized in any way that `DiagnosticBuilder` supports. It is a bit of a hack, but it works for now and should work for you.

**Comment by jsgf** [comment]
@estebank I think the diagnostics are OK now for a first pass. I'd like to revisit this with some way of directly referencing the source of the stray dependency which can use that mechanism.

**Comment by est31** [comment]
rustc only sees a limited part of the world, and any rustc-based lint will always be unsuited for rustc-with-Cargo use cases. Any good lint has to reside in Cargo, as outlined [in the issue](https://github.com/rust-lang/rust/issues/57274#issuecomment-629956427). Yes, one can make the `--extern` params for dev dependencies special and let the lint auto ignore them, but a good lint would also find unused dev dependencies, which only works if all examples/tests are compiled and usage info from all of them combined. Rustc only sees one example/doctest/etc at a time.

That being said, temporarily, until Cargo gets a native lint, it could pass `-Wunused-crate-deps` by default for `build.rs`, where the lint would work perfectly.

Another place where the lint works perfectly are [rustc-with-buck use cases](https://github.com/rust-lang/rust/issues/57274#issuecomment-624955717). This use case creates permanent legitimization of the feature inside rustc.

In the ideal end situation, the lint would be implemented twice: once in rustc, for tools like buck, and then in tools like Cargo that have a more complicated situation of extern crates.

**Comment by est31** [comment]
Actually disregard that, I think it's better if the lint will be enabled eventually, just some deps are suppressed. For cargo that would help find unused proper deps and unused build deps. The dev-deps can be checked by cargo's implementation, but it would only have to cover the dev-deps, nothing more.

**Comment by jsgf** [comment]
@est31 Implementing the lint in Cargo seems pretty awkward, since Cargo can't directly tell what crates the source is accessing. Using the deps output seems pretty brittle. I definitely think that its a rustc-level task, which is helped by extra information from the build system.

The Cargo limitation is that it has a pretty coarse model of dependencies - broadly only at the package scope, `build-` and `dev-dependencies` notwithstanding. It would be nice if we can refine the dependency model to something more fine grained (eg bin targets get separate deps from libraries in the same package). The interaction with this lint is just another fallout of that.

**Comment by est31** [comment]
> Using the deps output seems pretty brittle.

Note that the file format was *made* for compilers to tell build tools which files it uses. So we'd be using it for the purpose it was made for. There are a few gotchas that make it a bit brittle, indeed (especially in cargo-udeps), but if it ever turns out to be too brittle, one could think about creating json files or printing stuff to stdout (cargo gets json stuff from rustc already now).

> It would be nice if we can refine the dependency model to something more fine grained (eg bin targets get separate deps from libraries in the same package).

While there is a trend for more refinement, and I do support introduction of separate binary dependencies, I doubt that the refinement will reach levels where each separate test in the `tests/` directory or each separate example in the `examples/` directory has its own set of dependencies. If almost all of your tests except one use a specific crate, what should you do? List it in all tests?

**Comment by jsgf** [comment]
OK, I backported this to 1.43 and deployed it on our internal compiler, so I now have some non-toy scale experience with this.

TL;DR: It is very effective - there are a lot of latent unused dependencies in our codebase, and this is a useful tool for finding them.

I wrote a prototype script to try and take these lint outputs from rustc and automatically apply fixes based on them. I found two key problems with doing this:
1. While the crate name and rule name are often the same, it's not always the case. In Cargo terms, the package name and the crate name are not always the same, so you can't always correlate the compiler's diagnostic with an actual dependency.
2. The build system conflates test and non-test build output, and you can't distinguish them from the diagnostic alone. The problem is that the diagnostic identifies the source of the problem by the path to the source file, but the one source file could be used for a number of different targets. If you're building them together then the diagnostics will get jumbled without being able to tell them apart. Ultimately I think this is a build system problem - it should always be possible to tell which diagnostic came from which target, and in Buck builds it's common to be building the test and non-test variants of a crate the same time in parallel. I think Cargo has the same issue, though test builds tend to be distinct from non-test ones (but you could have a crate being built as a test and as a dependency of another test).

I think 1. is best solved by having a way for the build system to pass in the location of each dependency, such as the [`--extern-loc`](https://github.com/rust-lang/rust/issues/57274#issuecomment-631303573) option I proposed. For Buck this would encode a (target, dependency) tuple for each extern crate so that it can be reported back in the diagnostic. (I suspect Bazel would be similar.) For Cargo it would probably be (Cargo.toml file, line).

2. is also best solved in the build system, by having a mechanism to make sure every piece of diagnostic output is unambiguously tied to a specific build target - eg, abstractly, "this comes from a test build of crate foo", "normal build of foo", "build script for foo", etc. In Buck's case, it would be the specific rule name. But a short term fix might be to include a `test` field in the json diagnostics to indicate whether it was a `--test` build. This would allow an auto-fix tool to tell whether an unused dep is really completely unused, or a test/dev dependency. (Right now I have to run 2 passes - one which makes unused deps into test deps, and then if they're still unused, remove them altogether.)

Despite all this, the script was capable of automatically removing the vast majority of unused deps - either entirely, or moving deps only used for tests into `test_deps` (ie dev-dependencies).

File: src/test/ui/unused-crate-deps/unused-aliases.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Warn about unused aliased for the crate
+
+// edition:2018
+// check-pass
+// aux-crate:bar=bar.rs
+// aux-crate:barbar=bar.rs
+// no-prefer-dynamic
+// ignore-tidy-linelength
```
Comment (petrochenkov) [inline src/test/ui/unused-crate-deps/unused-aliases.rs:8]: Again, better to remove.
This and all other `ignore-tidy-linelength`s as well.

File: src/test/ui/unused-crate-deps/warn-attr.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Check for unused crate dep, no path
+
+// edition:2018
+// check-pass
+// aux-crate:bar=bar.rs
+// no-prefer-dynamic
```
Comment (petrochenkov) [inline src/test/ui/unused-crate-deps/warn-attr.rs:6]: Unnecessary?
(This and other `no-prefer-dynamic`s.)

Comment (petrochenkov) [reply src/test/ui/unused-crate-deps/warn-attr.rs:6]: If the `--crate-type lib` and `no-prefer-dynamic` were used to test the lint in different configurations, then it's better to merge all these tests into one.
We know that the lint is entirely unaffected by the crate types and linking types, and testing time is a finite resource.

File: src/librustc_session/lint/builtin.rs [hunk 3]
```diff
@@ -71,6 +71,12 @@ declare_lint! {
     "extern crates that are never used"
 }
 
+declare_lint! {
+    pub UNUSED_CRATE_DEPS,
```
Comment (petrochenkov) [inline src/librustc_session/lint/builtin.rs:75]:
```suggestion
    pub UNUSED_EXTERN_OPTIONS,
```
https://github.com/rust-lang/rust/pull/69203 used `unused_extern_options` because that's what the lint does exactly (at least from rustc's point of view).
Crate dependencies may be introduced without the options (with `extern crate`) and unused `--extern` options may not even point to existing crates.

File: src/librustc_metadata/creader.rs [hunk 4]
```diff
@@ -839,6 +879,46 @@ impl<'a> CrateLoader<'a> {
...
+        });
+
+        // Make a point span rather than covering the whole file
+        let spandata = krate.span.data();
+        let span = spandata.with_hi(spandata.lo);
+        // Complain about anything left over
+        for ext in externs {
+            self.sess.parse_sess.buffer_lint(
+                lint::builtin::UNUSED_CRATE_DEPS,
+                span,
```
Comment (petrochenkov) [inline src/librustc_metadata/creader.rs:911]:
```suggestion
                DUMMY_SP,
```
Dummy span actually points to `main_file.rs:0:0` so using the crate span should be unnecessary.

File: src/librustc_metadata/creader.rs [hunk 5]
```diff
@@ -839,6 +879,46 @@ impl<'a> CrateLoader<'a> {
...
+                ast::CRATE_NODE_ID,
+                &format!(
+                    "External crate `{}` unused in `{}`. Remove the dependency or add `use {} as _;`.",
```
Comment (petrochenkov) [inline src/librustc_metadata/creader.rs:914]:
```suggestion
                    "external crate `{}` unused in `{}`: remove the dependency or add `use {} as _;`",
```
(Error and warning messages follow some common style.)

**Comment by petrochenkov** [comment]
The logic in `creader.rs` looks more complicated than it needs to be.
There should be a single point at which we can insert names into a `HashSet<Symbol>` as used and then compare the collected set from the command line.
I'll try to find that point a bit later today.

**Comment by petrochenkov** [comment]
> The logic in `creader.rs` looks more complicated than it needs to be.
> There should be a single point at which we can insert names into a `HashSet<Symbol>` as used and then compare the collected set from the command line.
> I'll try to find that point a bit later today.

Here it is - https://github.com/petrochenkov/rust/commit/e649e8ccfc45b9d75142c5d02775278e65c5f685.

File: src/librustc_metadata/creader.rs (hunk 4 above)
Comment (jsgf) [reply src/librustc_metadata/creader.rs:911]: I tried this initially, but using it suppresses any kind of location reporting at all:
```
warning: External crate `bar` unused in `warn_cmdline`. Remove the dependency or add `use bar as _;`.
   |
   = note: requested on the command line with `-W unused-crate-deps`

warning: 1 warning emitted

```

File: src/librustc_session/lint/builtin.rs (hunk 3 above)
Comment (jsgf) [reply src/librustc_session/lint/builtin.rs:75]: I see what you mean, but referencing `--extern` is a bit obscure. `--extern` is not typically seen or used by end users, and they may not even be aware of them. Talking about "dependencies" is a bit higher level and more directly what the programmer is going to be using.

I'm not sure what you mean by `extern crate`. From my point of view, I see this lint as being responsible for highlighting a mismatch between the build system dependencies (as manifest by `--extern`) and what the source says - ie, does the source reference an external dependency? By this standard `extern crate` doesn't count because it's necessarily part of the sources - and we have `UNUSED_EXTERN_CRATES` for flagging those. Also I don't think it can actually add a new dependency can it? It's just a reference to one in the same manner as `use`, and so needs a corresponding `--extern`? (I guess std, core, etc excepted, but I think its reasonable for them to be exempt from this lint too.)

While `--extern` can refer to non-existent crates, and I think its useful to report that, it should never happen in practice unless the build system is broken, or you're manually invoking `rustc`.

File: src/test/ui/unused-crate-deps/warn-attr.rs (hunk 2 above)
Comment (jsgf) [reply src/test/ui/unused-crate-deps/warn-attr.rs:6]: OK, agreed.

One other thing I couldn't set up in the test framework was transitive dependencies - I wanted to make a test verifying that if A depends on B and B on C, but C is unused, A gets no warnings. Can that be done? It didn't look like aux crates could have aux crates themselves.

File: src/test/ui/unused-crate-deps/unused-aliases-dylib.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Warn about unused aliased for the crate
+
+// edition:2018
+// check-pass
+// aux-crate:bar=bar.rs
+// aux-crate:barbar=bar.rs
+// ignore-tidy-linelength
```
[the remark this answers is not in the document]
Comment (jsgf) [reply src/test/ui/unused-crate-deps/unused-aliases-dylib.rs:7]: I wasn't sure of what the actual matching rules are. Is it prefix? Any substring?

File: src/test/ui/unused-crate-deps/libfib.rs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+// check-pass
+// aux-crate:bar=bar.rs
+// compile-flags:--crate-type lib -Wunused-crate-deps
```
[the remark this answers is not in the document]
Comment (jsgf) [reply src/test/ui/unused-crate-deps/libfib.rs:3]: In this case its to make sure the lint works for library crates, not just bin. I'll add the comment.

File: src/test/ui/unused-crate-deps/unused-aliases-dylib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/test/ui/unused-crate-deps/unused-aliases-dylib.rs:7]: Any substring.

File: src/test/ui/unused-crate-deps/warn-attr.rs (hunk 2 above)
Comment (petrochenkov) [reply src/test/ui/unused-crate-deps/warn-attr.rs:6]: Looks like some UI tests work this way, e.g. `src\test\ui\static\auxiliary\static-priv-by-default.rs` depends on `src\test\ui\static\auxiliary\static_priv_by_default.rs`.

**Comment by jsgf** [comment]
@petrochenkov Yes, that's certainly simpler. I'll double-check it.

**Comment by jsgf** [comment]
Updated with review comment fixes.

File: src/test/ui/unused-crate-deps/libfib.rs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// Test warnings for a library crate
+
+// check-pass
+// aux-crate:bar=bar.rs
+// compile-flags:--crate-type lib -Wunused-crate-deps
+
+pub fn fib(n: u32) -> Vec<u32> {
+//~^ external crate `bar` unused in
```
Comment (petrochenkov) [inline src/test/ui/unused-crate-deps/libfib.rs:8]:
```suggestion
//~^ WARNING external crate `bar` unused in
```

File: src/librustc_session/lint/builtin.rs (hunk 3 above)
Comment (petrochenkov) [reply src/librustc_session/lint/builtin.rs:75]: Ok, in that case `unused_crate_deps` should probably be renamed to `unused_crate_dependencies` to avoid jargon and match the `Cargo.toml` section where they are specified.

**Comment by petrochenkov** [comment]
r=me after squashing the commits

**Comment by jsgf** [comment]
Squashed and rebased.

**Comment by jsgf** [comment]
Found existing method `Span::shrink_to_lo` instead of open-coding it.

**Comment by retep998** [comment]
This broke winapi CI
https://github.com/retep998/winapi-rs/runs/717012925

**Comment by petrochenkov** [comment]
@retep998 
Fixed in https://github.com/rust-lang/rust/pull/72702.
